### PR TITLE
Add a context manager to SeqIO.parse and avoid side effects

### PR DESCRIPTION
**Pull request by mdehoon**
Same as #4280 , updated for current master

The AlignIO stuff is ugly, but I don't see a better solution for that and anyway AlignIO may be on its way out.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Tests/test_SeqIO.py [hunk 1]
```diff
@@ -61,12 +62,16 @@ def get_mode(cls, fmt):
...
-                pass
+            with warnings.catch_warnings():
+                warnings.simplefilter("error")
+                try:
+                    SeqIO.read(stream, fmt)
+                except StreamModeError:
+                    continue
+                except BiopythonDeprecationWarning:  # uniprot-xml
+                    continue
+                except ValueError:  # SeqIO.read will complain that the stream is empty
```
Comment (peterjc) [inline Tests/test_SeqIO.py:73]: Can you confirm the exception message matches here. Possibly one of the unittest assert raises methods will cover this.

**Review by peterjc — CHANGES_REQUESTED** [review]
We should acknowledge @BabaYaga1221 in the commit message as the original or co-author. That would be easy to do in GitHub via the squash-and-merge button as you can edit the commit message (the default is rarely very sensible).

**Comment by peterjc** [comment]
Would you welcome automated edits to pull requests to apply black style? This would mean you may have to do a "git pull" before continuing local work though...

[I thought I had an open pull request to do this, I can make one if not]

File: Tests/test_SeqIO.py (hunk 1 above)
Comment (mdehoon) [reply Tests/test_SeqIO.py]: Done. Note that this is not a test, but a helper function to figure out the right mode (text or binary) to use.

**Pull request merged**